### PR TITLE
Improve traffic-managers podCIDR discovery

### PR DESCRIPTION
**Pull request by thallgren**
## Description

This PR contains two improvements to the pod CIDR discovery mechanism.
1. Performance improved by replacing the kates watchers with standard k8s APIs (#2084).
2. Make pod discovery strategy configurable in the Helm chart (#2023).

## Checklist

 - [x] I made sure to update `./CHANGELOG.md`.
 - [x] My change is adequately tested.

### CODE REVIEW

File: charts/telepresence/values.yaml [hunk 1]
```diff
@@ -70,6 +70,20 @@ grpc: {}
...
+# podCIDRs is the verbatim list of CIDRs used when the podCIDRStrategy is set to environment
+podCIDRs: []
+
+# podCIDRStrategy controls what strategy the traffic-manager will use for finding out what
+# CIDRs the cluster is using for its pods. Valid values are:
+#
+#  nodePodCIDRs extract CIDRs from the podCIDR and podCIDRs field of the Node Spec.
+#  coverPodIPs  extract IPs from the podIP and podIPs field of the Pod Status and compute the CIDRs needed to cover those IPs.
+#  environment  use CIDRs listed in the space separated POD_CIDRS environment variable verbatim.
+#  auto         first tru nodePodCIDRs and if that fails, tru coverPodIPs
```
Comment (josecv) [inline charts/telepresence/values.yaml]: typo -- it says `tru` instead of `try`

File: cmd/traffic/cmd/manager/internal/cluster/nodewatcher.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+	"github.com/datawire/dlib/dlog"
+	"github.com/telepresenceio/telepresence/v2/pkg/subnet"
+)
+
+type nodeWatcher struct {
+	lister   licorev1.NodeLister
+	informer cache.SharedIndexInformer
+	subnets  []*net.IPNet
+	changed  time.Time
+	lock     sync.Mutex // Protects all access to ipsMap
```
Comment (josecv) [inline cmd/traffic/cmd/manager/internal/cluster/nodewatcher.go]: should this read `Protects all access to subnets`?

File: cmd/traffic/cmd/manager/internal/cluster/nodewatcher.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+	if changed {
+		dlog.Infof(ctx, "Found %d subnets", len(w.subnets))
+		w.changed = time.Now()
+	} else {
+		dlog.Info(ctx, "No subnets found")
+	}
+	return changed
+}
+
+func (w *nodeWatcher) watch(ctx context.Context) {
```
Comment (josecv) [inline cmd/traffic/cmd/manager/internal/cluster/nodewatcher.go]: maybe i'm just confused but it seems like watch is never invoked?

File: cmd/traffic/cmd/manager/internal/cluster/nodewatcher.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+		}
+	}
+	if changed {
+		dlog.Infof(ctx, "Found %d subnets", len(w.subnets))
+		w.changed = time.Now()
+	} else {
+		dlog.Info(ctx, "No subnets found")
+	}
+	return changed
+}
```
Comment (josecv) [inline cmd/traffic/cmd/manager/internal/cluster/nodewatcher.go]: it seems like this function combines initialization with a precondition check (i.e. "Are we actually able to watch subnets from nodes?") -- that's okay, but calling it twice in a row will make the second call return `false`, no? Maybe we should short-circuit return if `w.subnets` is already populated -- after all, if they're there then getting them is definitely viable

File: cmd/traffic/cmd/manager/internal/cluster/podwatcher.go [hunk 5]
```diff
@@ -0,0 +1,247 @@
+package cluster
```
Comment (josecv) [inline cmd/traffic/cmd/manager/internal/cluster/podwatcher.go:1]: same comments apply here as for the nodewatcher :)

**Review by josecv — COMMENTED** [review]
looks good but i'm somewhat confused about the `watch` function --just wanna make sure that it actually _is_ being invoked, otherwise none of this will work 🙃

**Comment by thallgren** [comment]
The `watch` function was never called, but that only meant that future updates to podCIDRs after the initial scan where undetected. The initial scan happened anyway (in `viable`). Not good of course, but not a case of "none of this will work" either.

File: charts/telepresence/values.yaml [hunk 6]
```diff
@@ -79,7 +79,7 @@ podCIDRs: []
 #  nodePodCIDRs extract CIDRs from the podCIDR and podCIDRs field of the Node Spec.
 #  coverPodIPs  extract IPs from the podIP and podIPs field of the Pod Status and compute the CIDRs needed to cover those IPs.
 #  environment  use CIDRs listed in the space separated POD_CIDRS environment variable verbatim.
-#  auto         first tru nodePodCIDRs and if that fails, tru coverPodIPs
+#  auto         first try nodePodCIDRs and if that fails, tru coverPodIPs
```
Comment (josecv) [inline charts/telepresence/values.yaml:82]: still says `tru coverPodIPs` :^)

**Review by josecv — APPROVED** [review]

**Pull request merged**